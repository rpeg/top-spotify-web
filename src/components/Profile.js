import React, { useState, useEffect } from "react";
import axios from "axios";

import { isEmpty } from "lodash";

const Profile = () => {
  const [profile, setProfile] = useState({});

  useEffect(() => {
    const fetchProfile = async () => {
      const result = await axios
        .get("http://localhost:3000/my-profile")
        .catch(err => console.log(err));

      setProfile(result.data);
    };

    fetchProfile();
  }, []);

  return (
    !isEmpty(profile) && (
      <div>
        <img
          src={profile.images[0].url}
          alt={profile.id}
          style={{
            objectFit: "cover",
            borderRadius: "50%"
          }}
        />
        <h4>{profile.id}</h4>
      </div>
    )
  );
};

export default Profile;
