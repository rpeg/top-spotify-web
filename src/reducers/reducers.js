import { combineReducers } from 'redux';
import {
  SET_USER,
  SET_TIME_RANGE,
  TimeRanges,
  REQUEST_ARTISTS,
  RECEIVE_ARTISTS,
  REQUEST_TRACKS,
  RECEIVE_TRACKS,
  REQUEST_FEATURES,
  RECEIVE_FEATURES,
  SET_ARTIST_COUNT,
  SET_TRACK_COUNT,
  SET_GENRE_COUNT,
} from '../constants/constants';

const { LONG } = TimeRanges;

function user(state = {}, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, user: action.user };
    default:
      return state;
  }
}

function timeRange(state = LONG, action) {
  switch (action.type) {
    case SET_TIME_RANGE:
      return { ...state, timeRange: action.timeRange };
    default:
      return state;
  }
}

function artistCount(state = 10, action) {
  switch (action.type) {
    case SET_ARTIST_COUNT:
      return { ...state, artistCount: action.artistCount };
    default:
      return state;
  }
}

function trackCount(state = 15, action) {
  switch (action.type) {
    case SET_TRACK_COUNT:
      return { ...state, trackCount: action.trackCount };
    default:
      return state;
  }
}

function genreCount(state = 40, action) {
  switch (action.type) {
    case SET_GENRE_COUNT:
      return { ...state, genreCount: action.genreCount };
    default:
      return state;
  }
}

function artists(state = {}, action) {
  switch (action.type) {
    case REQUEST_ARTISTS:
      return { ...state, isFetching: true };
    case RECEIVE_ARTISTS:
      return { ...state, isFetching: false, items: action.items };
    default:
      return state;
  }
}

function artistsByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_ARTISTS:
    case REQUEST_ARTISTS:
      return { ...state, [action.timeRange]: artists(state[action.timeRange], action) };
    default:
      return state;
  }
}

function tracks(state = {}, action) {
  switch (action.type) {
    case REQUEST_TRACKS:
      return { ...state, isFetching: true };
    case RECEIVE_TRACKS:
      return { ...state, isFetching: false, items: action.tracks };
    default:
      return state;
  }
}

function tracksByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_TRACKS:
    case REQUEST_TRACKS:
      return { ...state, [action.timeRange]: tracks(state[action.timeRange], action) };
    default:
      return state;
  }
}

function features(state = {}, action) {
  switch (action.type) {
    case REQUEST_FEATURES:
      return { ...state, isFetching: true };
    case RECEIVE_FEATURES:
      return { ...state, isFetching: false, items: action.features };
    default:
      return state;
  }
}

const rootReducer = combineReducers({
  user,
  timeRange,
  artistCount,
  trackCount,
  genreCount,
  artists,
  artistsByTimeRange,
  tracks,
  tracksByTimeRange,
  features,
});

export default rootReducer;
