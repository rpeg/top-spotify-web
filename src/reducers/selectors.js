export const selectHasClickedCreate = (state) => state.hasClickedCreate;
export const selectUser = (state) => state.user;
export const selectTimeRangeName = (state) => state.timeRangeName;
export const selectGenreCount = (state) => state.genreCount;
export const selectArtistCount = (state) => state.artistCount;
export const selectTrackCount = (state) => state.trackCount;
export const selectStatsOptions = (state) => state.statsOptions;
export const selectDisplayProfile = (state) => state.displayProfile;
export const selectOptimizeTracks = (state) => state.optimizeTracks;
export const selectArtists = (state) => state.artists;
export const selectArtistsByCurrentTimeRange = (state) => state.artistsByTimeRangeName[state.timeRangeName];
export const selectTracks = (state) => state.tracks;
export const selectTracksByCurrentTimeRange = (state) => state.tracksByTimeRangeName[state.timeRangeName];
export const selectFeatures = (state) => state.features;
export const selectFeaturesByCurrentTimeRange = (state) => state.featuresByTimeRangeName[state.timeRangeName];
