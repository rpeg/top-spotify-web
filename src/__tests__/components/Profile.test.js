import React from 'react';
import { shallow } from 'enzyme';

import Profile from '../../components/Profile';
import * as selectors from '../../reducers/selectors';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((fn) => fn()),
  useDispatch: () => jest.fn(),
}));

const setup = (displayProfile = false) => {
  jest.spyOn(selectors, 'selectDisplayProfile').mockReturnValue(displayProfile);
};

describe('Profile', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should not log errors in console', () => {
    setup(true);
    const wrapper = shallow(<Profile />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders correctly with !displayProfile', () => {
    setup(false);
    const user = { id: 'abc' };

    const wrapper = shallow(<Profile user={user} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').exists()).toBe(false);
  });

  it('renders correctly with displayProfile', () => {
    setup(true);

    const user = { id: 'abc' };

    const wrapper = shallow(<Profile user={user} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').exists()).toBe(true);
  });

  it('renders correctly without user', () => {
    setup(true);

    const wrapper = shallow(<Profile />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').exists()).toBe(false);
  });

  it('renders correctly with user', () => {
    setup(true);

    const user = { id: 'abc' };

    const wrapper = shallow(<Profile user={user} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').exists()).toBe(true);
  });
});
