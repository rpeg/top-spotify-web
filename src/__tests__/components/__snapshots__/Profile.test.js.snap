// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly with !displayProfile 1`] = `""`;

exports[`Profile renders correctly with displayProfile 1`] = `
<span
  style={
    Object {
      "alignItems": "center",
      "display": "inline-flex",
      "marginLeft": "auto",
    }
  }
>
  <p
    className="text-m"
    style={
      Object {
        "margin": "0 0.5em 0 0",
      }
    }
  >
    <b>
      abc
    </b>
  </p>
  <img
    alt="abc"
    height="50px"
    style={
      Object {
        "borderRadius": "50%",
        "objectFit": "cover",
      }
    }
  />
</span>
`;

exports[`Profile renders correctly with user 1`] = `
<span
  style={
    Object {
      "alignItems": "center",
      "display": "inline-flex",
      "marginLeft": "auto",
    }
  }
>
  <p
    className="text-m"
    style={
      Object {
        "margin": "0 0.5em 0 0",
      }
    }
  >
    <b>
      abc
    </b>
  </p>
  <img
    alt="abc"
    height="50px"
    style={
      Object {
        "borderRadius": "50%",
        "objectFit": "cover",
      }
    }
  />
</span>
`;

exports[`Profile renders correctly without user 1`] = `""`;
